File: salesos/frontend/src/lib/api/promptLibrary.ts
/**
 * Studio Prompt Library HTTP (STORY-12-01 / FE-S12-01).
 * Tip in-memory CAP-089. feature_ai_copilot remains False.
 * No live LLM / RAG GO. Not Production GO.
 */
import api from "./client";

const BASE = "/api/v1/studio/prompt-library";

function tenantHeaders(tenantId: string) {
  return { "X-Tenant-Id": tenantId };
}

export interface PromptVersionRecord {
  version: string;
  template: string;
  system: string;
  changelog: string;
  created_at?: string;
}

export interface PromptLibraryEntry {
  id: string;
  tenant_id: string;
  name: string;
  key: string;
  active_version: string;
  versions: PromptVersionRecord[];
  domain: string;
  category: string;
  schema_version: number;
  created_at?: string;
  updated_at?: string;
  version_count: number;
}

export interface PromptLibraryMeta {
  object: string;
  capability: string;
  extends: string;
  operations: string[];
  feature_ai_copilot: boolean;
  honesty: string;
}

export interface PromptCreateBody {
  name: string;
  key: string;
  template: string;
  system?: string;
  version?: string;
  changelog?: string;
  domain?: string;
  category?: string;
  id?: string;
}

export interface PromptVersionBody {
  template: string;
  version: string;
  system?: string;
  changelog?: string;
  activate?: boolean;
}

export interface PromptRollbackBody {
  version: string;
}

export interface PromptMetaPatchBody {
  name?: string;
  domain?: string;
  category?: string;
}

export async function getPromptLibraryMeta(
  tenantId: string,
): Promise<PromptLibraryMeta> {
  const resp = await api.get<PromptLibraryMeta>(`${BASE}/meta`, {
    headers: tenantHeaders(tenantId),
  });
  return resp.data;
}

export async function listPromptLibrary(
  tenantId: string,
): Promise<PromptLibraryEntry[]> {
  const resp = await api.get<PromptLibraryEntry[]>(BASE, {
    headers: tenantHeaders(tenantId),
  });
  return resp.data;
}

export async function getPromptLibraryEntry(
  tenantId: string,
  entryId: string,
): Promise<PromptLibraryEntry> {
  const resp = await api.get<PromptLibraryEntry>(
    `${BASE}/${encodeURIComponent(entryId)}`,
    { headers: tenantHeaders(tenantId) },
  );
  return resp.data;
}

export async function createPromptLibraryEntry(
  tenantId: string,
  body: PromptCreateBody,
): Promise<PromptLibraryEntry> {
  const resp = await api.post<PromptLibraryEntry>(BASE, body, {
    headers: tenantHeaders(tenantId),
  });
  return resp.data;
}

export async function patchPromptLibraryMeta(
  tenantId: string,
  entryId: string,
  body: PromptMetaPatchBody,
): Promise<PromptLibraryEntry> {
  const resp = await api.patch<PromptLibraryEntry>(
    `${BASE}/${encodeURIComponent(entryId)}`,
    body,
    { headers: tenantHeaders(tenantId) },
  );
  return resp.data;
}

export async function addPromptLibraryVersion(
  tenantId: string,
  entryId: string,
  body: PromptVersionBody,
): Promise<PromptLibraryEntry> {
  const resp = await api.post<PromptLibraryEntry>(
    `${BASE}/${encodeURIComponent(entryId)}/versions`,
    body,
    { headers: tenantHeaders(tenantId) },
  );
  return resp.data;
}

export async function rollbackPromptLibrary(
  tenantId: string,
  entryId: string,
  body: PromptRollbackBody,
): Promise<PromptLibraryEntry> {
  const resp = await api.post<PromptLibraryEntry>(
    `${BASE}/${encodeURIComponent(entryId)}/rollback`,
    body,
    { headers: tenantHeaders(tenantId) },
  );
  return resp.data;
}

export async function deletePromptLibraryEntry(
  tenantId: string,
  entryId: string,
): Promise<{ deleted: boolean; id: string }> {
  const resp = await api.delete<{ deleted: boolean; id: string }>(
    `${BASE}/${encodeURIComponent(entryId)}`,
    { headers: tenantHeaders(tenantId) },
  );
  return resp.data;
}

File: salesos/backend/app/modules/tenant_studio/prompt_library.py
"""STORY-12-01 — CAP-089 Prompt Library models (tenant-facing, extends CAP-023).

Tenant CRUD + versioning + rollback over Prompt Registry-shaped entries.
Not Production GO. DEC-085 untouched. No Alembic / FORCE RLS.
feature_ai_copilot remains False. No live LLM / RAG GO.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any


class PromptLibraryError(ValueError):
    """Invalid prompt library entry or version operation."""


@dataclass(frozen=True)
class PromptVersionRecord:
    """Immutable version snapshot for a tenant prompt."""

    version: str
    template: str
    system: str = ""
    changelog: str = ""
    created_at: str = ""

    def as_dict(self) -> dict[str, Any]:
        return {
            "version": self.version,
            "template": self.template,
            "system": self.system,
            "changelog": self.changelog,
            "created_at": self.created_at,
        }


@dataclass
class PromptLibraryEntry:
    """Tenant-owned prompt library entry (CAP-089 / CAP-023-shaped)."""

    id: str
    tenant_id: str
    name: str
    key: str
    active_version: str
    versions: list[PromptVersionRecord] = field(default_factory=list)
    domain: str = "gtm"
    category: str = "general"
    schema_version: int = 1
    created_at: str = ""
    updated_at: str = ""

    def as_dict(self) -> dict[str, Any]:
        return {
            "id": self.id,
            "tenant_id": self.tenant_id,
            "name": self.name,
            "key": self.key,
            "active_version": self.active_version,
            "versions": [v.as_dict() for v in self.versions],
            "domain": self.domain,
            "category": self.category,
            "schema_version": self.schema_version,
            "created_at": self.created_at,
            "updated_at": self.updated_at,
            "version_count": len(self.versions),
        }

    def active_record(self) -> PromptVersionRecord | None:
        for v in self.versions:
            if v.version == self.active_version:
                return v
        return self.versions[-1] if self.versions else None


def normalize_key(key: str) -> str:
    k = (key or "").strip().lower().replace(" ", ".")
    if not k:
        raise PromptLibraryError("key required")
    if not all(c.isalnum() or c in "._-" for c in k):
        raise PromptLibraryError("key must be alphanumeric with ._-")
    return k


def normalize_version(version: str) -> str:
    v = (version or "").strip()
    if not v:
        raise PromptLibraryError("version required")
    if len(v) > 32:
        raise PromptLibraryError("version too long")
    return v


def normalize_template(template: str) -> str:
    t = (template or "").strip()
    if not t:
        raise PromptLibraryError("template required")
    if len(t) > 20000:
        raise PromptLibraryError("template too long")
    return t
